bauble_macro_util/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+
+            let length = fields.len();
+            if length != 0 {
+                ::std::result::Result::Err(::bauble::DeserializeError::WrongTupleLength {
+                    expected: #field_count,
+                    found: #val_count + length,
+                    ty: self_type_info.clone(),
+                    span,
+                })?
+            }
+        },
+        false => quote! {
+            let values = (#( #values, )*);
+
+            if let ::std::option::Option::Some((field, _)) = fields.into_iter().next() {
+                ::std::result::Result::Err(::bauble::DeserializeError::UnexpectedField {
+                    field,
+                    ty: self_type_info.clone(),
+                })?
+            }
+        },
+    };
+
+    // Generate code that evaluates each field
+    let fields = fields.iter().map(|field| {
+        let ident = &field.name;
+        let default = Ident::new(&format!("default_{ident}"), Span::call_site());
+        match &field.ty {
+            FieldTy::Val {
+                default: Some(_),
+                index,
+                ..
+            } => quote! {
+                #ident: match values.#index {
+                    Some(value) => ::bauble::BaubleAllocator::validate(
+                        allocator,
+                        ::bauble::FromBauble::from_bauble(value, allocator)?,
+                    )?,
+                    None => #default(),
+                }
+            },
+            FieldTy::Val {
+                default: None,
+                index,
+                ..
+            } => quote! {
+                #ident: ::bauble::BaubleAllocator::validate(
+                    allocator,
+                    ::bauble::FromBauble::from_bauble(values.#index, allocator)?
+                )?
+            },
+            FieldTy::AsDefault { .. } => quote! { #ident: #default() },
+        }
+    });
+
+    // Assemble the deserialization code, including a check for unexpected attributes
+    quote! {
+        #( #defaults )*
+
+        #values
+
+        if let ::std::option::Option::Some((attribute, _)) = attributes.into_iter().next() {
+            ::std::result::Result::Err(::bauble::DeserializeError::UnexpectedAttribute {
+                attribute,
+                ty: self_type_info.clone(),
+            })?
+        }
+
+        unsafe {
+            ::bauble::BaubleAllocator::wrap(
+                allocator,
+                #ty {
+                    #( #fields, )*
+                }
+            )
+        }
+    }
+}
+
+// Generate code to deserialize a struct or variant. See `derive_fields` for more field docs.
+fn derive_struct(
+    ty_info: TypeInfo,
+    // struct / variant level attributes
+    attributes: Vec<Ident>,
+    fields: &Fields,
+) -> proc_macro2::TokenStream {
+    let mut tuple = false;
+
+    for attribute in attributes {
+        match attribute.to_string().as_str() {
+            "tuple" => tuple = true,
```

Comment:
Should probably have a check here to make sure there's only one "tuple" attribute. Currently, this code compiles:
```rs
#[derive(FromBauble)]
#[bauble(tuple)]
#[bauble(tuple)]
struct Test {
    x: i32,
    y: i32,
}
```